File: src/services/audit/detectors/ad/attack-paths/high/computer-takeover.ts
/**
 * Computer Takeover Path Detector
 *
 * Detects RBCD attack paths.
 * Computers with Resource-Based Constrained Delegation configured.
 */

import { ADUser, ADGroup, ADComputer } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { buildGroupMembershipGraph } from '../../../../../../utils/graph.util';

/**
 * Detect RBCD attack paths
 * Computers with Resource-Based Constrained Delegation configured
 */
export function detectPathComputerTakeover(
  computers: ADComputer[],
  users: ADUser[],
  groups: ADGroup[],
  includeDetails: boolean
): Finding {
  const graph = buildGroupMembershipGraph(users, groups, computers);

  // Find computers with RBCD configured
  const rbcdComputers = computers.filter((c) => {
    const rbcd = (c as any)['msDS-AllowedToActOnBehalfOfOtherIdentity'];
    return rbcd && rbcd !== '';
  });

  // Check which are privileged (DCs, admin workstations)
  const dcPattern = /^(dc|domain controller)/i;
  const affected = rbcdComputers.filter((c) => {
    const computerMemberOf = c['memberOf'] as string[] | undefined;
    const isDC = dcPattern.test(c.sAMAccountName || '') || computerMemberOf?.some((g: string) => g.includes('Domain Controllers'));
    const isPrivileged = graph.isPrivileged(c.dn);
    return isDC || isPrivileged;
  });

  return {
    type: 'PATH_COMPUTER_TAKEOVER',
    severity: 'high',
    category: 'attack-paths',
    title: 'RBCD Computer Takeover Path',
    description:
      'Privileged computers have Resource-Based Constrained Delegation configured. Attackers controlling the delegating principal can compromise these computers.',
    count: affected.length,
    affectedEntities: includeDetails
      ? affected.map((c) => ({
          type: 'computer' as const,
          id: c.dn,
          displayName: c.sAMAccountName || c.dn,
          sAMAccountName: c.sAMAccountName,
          dNSHostName: c.dNSHostName || null,
          operatingSystem: c.operatingSystem || null,
          operatingSystemVersion: c.operatingSystemVersion || null,
          whenCreated: null,
          whenChanged: null,
          lastLogon: null,
          pwdLastSet: null,
          enabled: c.enabled ?? true,
          userAccountControl: 0,
        }))
      : undefined,
    details: {
      totalRbcdComputers: rbcdComputers.length,
      privilegedRbcdComputers: affected.length,
      attackVector: 'Control delegating account → RBCD → Impersonate on target',
      mitigation: 'Remove RBCD from privileged computers, monitor msDS-AllowedToActOnBehalfOfOtherIdentity',
    },
  };
}

File: internal/audit/detectors/ad/attack-paths/high/computer-takeover.go
package high

import (
	"context"
	"regexp"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// ComputerTakeoverDetector detects RBCD attack paths
type ComputerTakeoverDetector struct {
	audit.BaseDetector
}

// NewComputerTakeoverDetector creates a new detector
func NewComputerTakeoverDetector() *ComputerTakeoverDetector {
	return &ComputerTakeoverDetector{
		BaseDetector: audit.NewBaseDetector("PATH_COMPUTER_TAKEOVER", audit.CategoryAttackPaths),
	}
}

var dcPattern = regexp.MustCompile(`(?i)^(dc|domain controller)`)

// Detect executes the detection
func (d *ComputerTakeoverDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var rbcdComputers []types.Computer
	var affected []types.Computer

	for _, c := range data.Computers {
		// Find computers with RBCD configured
		if len(c.AllowedToActOnBehalfOfOtherIdentity) == 0 {
			continue
		}
		rbcdComputers = append(rbcdComputers, c)

		// Check if privileged (DC or in Domain Controllers group)
		isDC := dcPattern.MatchString(c.SAMAccountName)
		for _, memberOf := range c.MemberOf {
			if strings.Contains(strings.ToLower(memberOf), "domain controllers") {
				isDC = true
				break
			}
		}

		if isDC || c.AdminCount {
			affected = append(affected, c)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "RBCD Computer Takeover Path",
		Description: "Privileged computers have Resource-Based Constrained Delegation configured. Attackers controlling the delegating principal can compromise these computers.",
		Count:       len(affected),
		Details: map[string]interface{}{
			"totalRbcdComputers":      len(rbcdComputers),
			"privilegedRbcdComputers": len(affected),
			"attackVector":            "Control delegating account → RBCD → Impersonate on target",
			"mitigation":              "Remove RBCD from privileged computers, monitor msDS-AllowedToActOnBehalfOfOtherIdentity",
		},
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedComputerEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewComputerTakeoverDetector())
}
